sourcecode/apis/contentauthor/app/Http/Controllers/Admin/AdminH5PDetailsController.php

```diff
@@ -185,6 +188,69 @@ public function contentHistory(H5PContent $content): View
         ]);
     }
 
+    public function libraryTranslation(H5PLibrary $library, string $locale): View
+    {
+        $libLang = H5PLibraryLanguage::where('library_id', $library->id)
+            ->where('language_code', $locale)
+            ->first();
+
+        return view('admin.library-upgrade.translation', [
+            'library' => $library,
+            'languageCode' => $locale,
+            'haveTranslation' => $libLang !== null,
+            'translationDb' => $libLang?->translation,
+            'translationFile' => Storage::disk()->get(
+                sprintf('libraries/%s/language/%s.json', $library->getFolderName(), $locale)
+            ),
+        ]);
+    }
+
+    public function libraryTranslationUpdate(AdminTranslationUpdateRequest $request, H5PLibrary $library, string $locale): View
+    {
+        $errorMsg = '';
+        $success = -1;
+        $input = $request->validated();
+
+        try {
+            if (array_key_exists('translationFile', $input) && $request->file('translationFile')->isValid()) {
+                $translation = $request->file('translationFile')->getContent();
+            } else {
+                $translation = $input['translation'];
+            }
+            if (empty($translation)) {
+                throw new Exception('Empty content');
+            }
+            json_decode($translation, flags: JSON_THROW_ON_ERROR);
+
+            $success = $library->languages()
+                ->where('language_code', $locale)
+                ->limit(1)
+                ->update(['translation' => $translation]);
+
+            if ($success === 0) {
+                $errorMsg = "No changes to save";
+            }
+        } catch (Exception $e) {
```

Comment:
What is this really meant to catch? There's a database query, which if fails, wouldn't be logged due to this. If it's meant to catch the JSON decoding, it should be narrowed down to that statement and/or the JsonException class.

Reply:
The content that was saved was the same as already in the database, the update didn't fail, it just didn't update any records

Reply:
Right, but this can still catch unrelated JSON/DB/Laravel exceptions. I believe a `goto` from the empty content check to viewing the template would be a better choice for the kind of code flow that's attempted here.